File: sample-videochat-webrtc/sample-videochat-webrtc/ViewControllers/Settings/BaseSettingsViewController/BaseSettingsViewController.m
//
//  BaseSettingsViewController.m
//  sample-videochat-webrtc-old
//
//  Copyright (c) 2017 QuickBlox. All rights reserved.
//

#import "BaseSettingsViewController.h"

#import "Settings.h"

#import "SettingCell.h"
#import "SettingSliderCell.h"
#import "SettingSwitchCell.h"

#import "SwitchItemModel.h"
#import "SliderItemModel.h"
#import "SettingsSectionModel.h"

@implementation BaseSettingsViewController

// MARK: Life cycle

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.settings = [Settings instance];
    self.selectedIndexes = [NSMutableDictionary dictionary];
    self.sections = [NSMutableDictionary dictionary];
    
    [self configure];
    [self registerNibs];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    [self applySettings];
}

// MARK: Public

- (SettingsSectionModel *)sectionWith:(NSUInteger)sectionType {
    
    NSString *title = [self titleForSection:sectionType];
    SettingsSectionModel *section = self.sections[title];
    return section;
}

- (NSIndexPath *)indexPathAtSection:(NSUInteger)section {
    
    NSString *key = [self titleForSection:section];
    NSIndexPath *indexPath = self.selectedIndexes[key];
    
    return indexPath;
}

- (BaseItemModel *)modelWithIndex:(NSUInteger)index section:(NSUInteger)section {
    
    SettingsSectionModel *sectionModel = [self sectionWith:section];
    
    if (sectionModel.items.count == 0) {
        return nil;
    }
    
    BaseItemModel *model = sectionModel.items[index];
    return model;
}

- (SettingsSectionModel *)addSectionWith:(NSUInteger)section items:(NSArray *(^)(NSString *sectionTitle))items {
    
    NSString *sectionTitle = [self titleForSection:section];
    SettingsSectionModel *sectionModel = [SettingsSectionModel sectionWithTitle:sectionTitle
                                                                          items:items(sectionTitle)];
    self.sections[sectionTitle] = sectionModel;
    
    return sectionModel;
}

- (void)selectSection:(NSUInteger)section index:(NSUInteger)index {
    
    if (index == NSNotFound) {
        index = 0;
    }
    
    NSString *sectionTitle = [self titleForSection:section];
    NSIndexPath *supportedFormatsIndexPath = [NSIndexPath indexPathForRow:index inSection:section];
    self.selectedIndexes[sectionTitle] = supportedFormatsIndexPath;
}

- (void)updateSelectionAtIndexPath:(NSIndexPath *)indexPath {
    
    NSString *key = [self titleForSection:indexPath.section];
    NSIndexPath *previosIndexPath = self.selectedIndexes[key];
    
    if ([indexPath compare:previosIndexPath] == NSOrderedSame) {
        return;
    }
    self.selectedIndexes[key] = indexPath.copy;
    
    [self.tableView reloadRowsAtIndexPaths:@[previosIndexPath, indexPath]
                          withRowAnimation:UITableViewRowAnimationFade];
}

// MARK: Override

- (void)configure {
    
    NSAssert(nil, @"Must be overriden in superclass.");
}

- (void)applySettings {
    
    NSAssert(nil, @"Must be overriden in superclass.");
}

- (NSString *)titleForSection:(NSUInteger)section {
    
    NSAssert(nil, @"Must be overriden in superclass.");
    
    return nil;
}

// MARK: UITableViewDataSource

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    
    SettingsSectionModel *sectionItem = [self sectionWith:section];
    return sectionItem.title;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    
    return self.sections.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    
    SettingsSectionModel *sectionItem = [self sectionWith:section];
    return sectionItem.items.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    SettingsSectionModel *sectionItem = [self sectionWith:indexPath.section];
    BaseItemModel *itemModel = sectionItem.items[indexPath.row];
    
    BaseSettingsCell *cell = [tableView dequeueReusableCellWithIdentifier:NSStringFromClass(itemModel.viewClass)];
    
    NSString *key = [self titleForSection:indexPath.section];
    
    NSIndexPath *selectedIndexPath = self.selectedIndexes[key];
    cell.accessoryType = [indexPath compare:selectedIndexPath] == NSOrderedSame ? UITableViewCellAccessoryCheckmark : UITableViewCellAccessoryNone;
    
    cell.delegate = self;
    cell.model = itemModel;
    
    return cell;
}

// MARK: SettingsCellDelegate

- (void)cell:(BaseSettingsCell *)cell didChageModel:(BaseItemModel *)model {
    
    NSAssert(nil, @"Required method of SettingsCellDelegate must be implemented in superclass.");
}

// Private:

- (void)registerNibs {
    
    [self.tableView registerNib:[SettingCell nib] forCellReuseIdentifier:[SettingCell identifier]];
    [self.tableView registerNib:[SettingSwitchCell nib] forCellReuseIdentifier:[SettingSwitchCell identifier]];
    [self.tableView registerNib:[SettingSliderCell nib] forCellReuseIdentifier:[SettingSliderCell identifier]];
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Settings/BaseSettingsViewController/BaseSettingsViewController.swift
//
//  BaseSettingsViewController.swift
//  sample-conference-videochat-swift
//
//  Created by Vladimir Nybozhinsky on 04.10.2018.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit

class BaseSettingsViewController: UITableViewController, SettingsCellDelegate {

    // MARK: Properties
    
    /**
     *  Settings storage.
     *
     *  @see Settings
     */
    var settings: Settings?
    /**
     *  Sections models.
     */
    var sections: [String: SettingsSectionModel] = [:]
    /**
     *  Selected indexes for each section.
     */
    var selectedIndexes: [String: IndexPath] = [:]

    // MARK: Life cycle
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        settings = Settings.instance
        selectedIndexes = [String: IndexPath]()
        sections = [String: SettingsSectionModel]()
        
        configure()
        registerNibs()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        applySettings()
    }

    // MARK: Public methods
    
    /**
     *  Settings section model for section index
     *
     *  @param sectionType section index
     *
     *  @return Settings section model
     */
    func section(with sectionType: Int) -> SettingsSectionModel? {
        
        let title = self.title(forSection: sectionType)
        let section: SettingsSectionModel? = sections[title ?? ""]
        return section
    }
    
    /**
     * Index path for section index.
     *
     *  @param section Section index
     *
     *  @return Index path for section index
     */
    func indexPath(atSection section: Int) -> IndexPath? {
        
        let key = title(forSection: section)
        let indexPath = selectedIndexes[key!]
        
        return indexPath
    }
    
    /**
     *  Model for section with index.
     *
     *  @param index model index
     *  @param section section index
     *
     *  @return model for section
     */
    func model(with index: Int, section: Int) -> BaseItemModel? {
        
        let sectionModel: SettingsSectionModel? = self.section(with: section)
        
        if sectionModel?.items.count == 0 {
            return nil
        }
        
        let model: BaseItemModel? = sectionModel?.items[index]
        return model
    }
    
    /**
     *  Add section with index and items.
     *
     *  @param section section index
     *  @param items items for section
     *
     *  @return settings section model
     */
    
    func addSection(with section: Int, items: @escaping (_ sectionTitle: String?) -> [BaseItemModel]) {
        
        let sectionTitle = title(forSection: section)
        let sectionModel = SettingsSectionModel.section(withTitle: sectionTitle, items: items(sectionTitle))
        sections[sectionTitle!] = sectionModel

    }
    
//    func addSection(with section: Int, items: @escaping (_ sectionTitle: String?) -> [Any]) -> SettingsSectionModel? {
//
//        let sectionTitle = title(forSection: section)
//        let sectionModel = SettingsSectionModel.section(withTitle: sectionTitle, items: items(sectionTitle) as? [BaseItemModel])
//        sections[sectionTitle!] = sectionModel
//
//        return sectionModel
//    }

    /**
     *  Select item at section.
     *
     *  @param section section index
     *  @param index item index
     */
    func selectSection(_ section: Int, index: Int) {
        var index = index
        
        if index == NSNotFound {
            index = 0
        }
        
        let sectionTitle = title(forSection: section)
        let supportedFormatsIndexPath = IndexPath(row: index, section: section)
        selectedIndexes[sectionTitle!] = supportedFormatsIndexPath
    }
    
    /**
     *  Update selection by selecting a new item and deselecting old one.
     *
     *  @param indexPath index path of requested item
     */
    func updateSelection(at indexPath: IndexPath?) {
        
        let key = title(forSection: (indexPath?.section)!)
        let previosIndexPath = selectedIndexes[key!]
        
        if let aPath = previosIndexPath {
            if ((indexPath?.compare(aPath)) != nil) {
                return
            }
        }
        if let aCopy = indexPath {
            selectedIndexes[key!] = aCopy
        }
        
        tableView.reloadRows(at: [previosIndexPath!, indexPath!], with: .fade)
    }
    
    // MARK: Override
    func configure() {
        assert(false, "Must be overriden in superclass.")
    }
    
    func applySettings() {
        assert(false, "Must be overriden in superclass.")
    }
    
    func title(forSection section: Int) -> String? {
        assert(false, "Must be overriden in superclass.")
        return nil
    }
    
    // MARK: UITableViewDataSource
    
    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        
        let sectionItem: SettingsSectionModel? = self.section(with: section)
        return sectionItem?.title
    }
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        
        return sections.count
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        
        let sectionItem: SettingsSectionModel? = self.section(with: section)
        return sectionItem?.items.count ?? 0
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let sectionItem: SettingsSectionModel? = section(with: indexPath.section)
        let itemModel: BaseItemModel? = sectionItem?.items[indexPath.row]
       
        debugPrint("itemModel? \(String(describing: itemModel?.title))")
       let cell = tableView.dequeueReusableCell(withIdentifier: NSStringFromClass((itemModel?.viewClass())!).components(separatedBy: ".").last!) as! BaseSettingsCell
            let key = title(forSection: indexPath.section)
        debugPrint(title(forSection: indexPath.section)!)
//            debugPrint("itemModel?.viewClass() \(String(describing: NSStringFromClass((itemModel?.viewClass())!).components(separatedBy: ".").last!))")
            let selectedIndexPath = selectedIndexes[key!]
//            if let aPath = selectedIndexPath {
//        cell.accessoryType = indexPath.compare(selectedIndexPath!) == .orderedSame ? UITableViewCell.AccessoryType.checkmark : UITableViewCell.AccessoryType.none
//            }
        
        cell.delegate = self
        cell.updateModel(itemModel!)
        
        return cell
    }
    
    // MARK: SettingsCellDelegate
    func cell(_ cell: BaseSettingsCell?, didChageModel model: BaseItemModel?) {
        
        assert(false, "Required method of SettingsCellDelegate must be implemented in superclass.")
    }
    
    // Private:
    func registerNibs() {

//        tableView.register(SettingCell.nib(), forCellReuseIdentifier: SettingCell.identifier()!)
//        tableView.register(SettingSwitchCell.nib(), forCellReuseIdentifier: SettingSwitchCell.identifier()!)
//        tableView.register(SettingSliderCell.nib(), forCellReuseIdentifier: SettingSliderCell.identifier()!)
        tableView.register(UINib(nibName: "SettingCell", bundle: nil), forCellReuseIdentifier: "SettingCell")
        tableView.register(UINib(nibName: "SettingSwitchCell", bundle: nil), forCellReuseIdentifier: "SettingSwitchCell")
        tableView.register(UINib(nibName: "SettingSliderCell", bundle: nil), forCellReuseIdentifier: "SettingSliderCell")
    }
}
